extensions/dialog/internal.m

```diff
@@ -0,0 +1,388 @@
+    }
+    
+    if (lua_isboolean(L, 3) && !lua_toboolean(L, 3)) {
+        [panel setCanChooseFiles:NO];
+    }
+    else
+    {
+        [panel setCanChooseFiles:YES];
+        
+    }
+    
+    if (lua_isboolean(L, 4) && lua_toboolean(L, 4)) {
+        [panel setCanChooseDirectories:YES];
+    }
+    else {
+        [panel setCanChooseDirectories:NO];
+    }
+    
+    if (lua_isboolean(L, 5) && lua_toboolean(L, 5)) {
+        [panel setAllowsMultipleSelection:YES];
+    }
+    else {
+        [panel setAllowsMultipleSelection:NO];
+    }
+
+    NSInteger clicked = [panel runModal];
+
+    int count = 1;
+    
+    if (clicked == NSFileHandlingPanelOKButton) {
+        for (NSURL *url in [panel URLs]) {
+            count = count + 1;
+            lua_pushstring(L,[[url absoluteString] UTF8String]);
+        }
+        count = count - 1;
+    }
+    else
+    {
+        lua_pushnil(L) ;
+    }
+    
+    return count ;
+}
+
+/// hs.dialog.webviewAlert(webview, callbackFn, message, informativeText, [buttonOne], [buttonTwo], [style]) -> string
+/// Function
+/// Displays a simple blocking dialog box using `NSAlert` in a `hs.webview`.
+///
+/// Parameters:
+///  * webview - The `hs.webview` to display the alert on.
+///  * callbackFn - The callback function that's called when a button is pressed.
+///  * message - The message text to display.
+///  * informativeText - The informative text to display.
+///  * [buttonOne] - An optional value for the first button as a string. Defaults to "OK".
+///  * [buttonTwo] - An optional value for the second button as a string. By default there is no second button.
+///  * [style] - An optional style of the dialog box as a string. Defaults to "NSWarningAlertStyle".
+///
+/// Returns:
+///  * The value of the button as a string.
+///
+/// Notes:
+///  * This alert is blocking (i.e. no other Lua code will be processed until the alert is closed).
+///  * The optional values must be entered in order (i.e. you can't supply `style` without also supplying `buttonOne` and `buttonTwo`).
+///  * [style] can be "NSWarningAlertStyle", "NSInformationalAlertStyle" or "NSCriticalAlertStyle". If something other than these string values is given, it will use "NSWarningAlertStyle".
+/*
+ EXAMPLE:
+ testWebview = hs.webview.newBrowser(hs.geometry.rect(250, 250, 250, 250)):show()
+ testCallbackFn = function(result) print("Callback Result: " .. result) end
+ hs.dialog.webviewAlert(testWebview, testCallbackFn, "Message", "Informative Text", "Button One", "Button Two", "NSCriticalAlertStyle")
+ */
+static int webviewAlert(lua_State *L) {
+    
+    NSString* defaultButton = @"OK";
+    
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, "hs.webview", LS_TFUNCTION, LS_TSTRING, LS_TSTRING, LS_TSTRING | LS_TOPTIONAL, LS_TSTRING | LS_TOPTIONAL, LS_TSTRING | LS_TOPTIONAL, LS_TBREAK];
+    
+    NSWindow *webview = [skin toNSObjectAtIndex:1];
+    //NSObject *callback = [skin toNSObjectAtIndex:2];
```

Comment:
~~~objc
    lua_pushvalue(L, 2) ; // copy the function to the top of the stack
    int callbackRef = [skin luaRef:refTable] ; // store what's at the top of the stack in the registry and save it's reference number
~~~

Reply:
Note `luaRef:` will pull off the top value of the stack, so the net effect of these two lines is to leave the stack of arguments as-is.